### PR TITLE
Add support for vendor extensions

### LINKED ISSUE #696: Add Support for Vendor Extensions
OpenAPI supports [vendor extensions ](https://swagger.io/docs/specification/openapi-extensions/). It would be very useful for Tsoa to add support for this so that developers can easily add properties to their spec files that aren't covered by the standard OpenAPI Specification.

## Sorting

- **I'm submitting a ...**

  - [ ] bug report
  - [x] feature request
  - [ ] support request

- I confirm that I
  - [x] used the [search](https://github.com/lukeautry/tsoa/search?type=Issues) to make sure that a similar issue hasn't already been submit

## Expected Behavior

```javascript
  @Extension({
    cmdlet: "Get-Endpoint",
    rolesAllowed: [
      "Administrator",
      "Audit Administrator"
    ],
    since: "1.0"
  })
  @Get()
  getEndpoints() {
    . . .
  }
```

would generate the following in the swagger/openapi spec file:

```json
"paths" : {
    "/endpoint" : {
      "get" : {
        "x-cmdlet" : "Get-Endpoint",
        "x-rolesAllowed" : [ "Administrator", "Audit Administrator"],
        "x-since" : "1.0"
      }
    }
 }
```

Note that in this example, other decorators (and properties in the spec file) were omitted for brevity.

## Current Behavior

Currently there is no way to provide vendor extensions. I suppose one might be able to use the spec merging feature to accomplish this, but you lose the benefit of being able to do this programmatically within your TypeScript source code.

## Possible Solution

My team originally forked this project to make these changes, but essentially a new Extension decorator was added to *src/decorators*, the *src/metadataGeneration/methodGenerator* was extended to look for the new Extension decorator, and the buildMethod() method in *src/swagger/specGenerator3* was extended to extract the values from the decorator and insert them into the spec file.   

## Context (Environment)

Version of the library: 3.0.8
Version of NodeJS: 12.16.1

- Confirm you were using yarn not npm: [x]

## Detailed Description

My team has a need to insert extra properties on our paths that aren't covered by the official OpenAPI spec. Other swagger/openapi libraries that we've used in other languages usually provide an *@extension* annotation for doing this. Our proposal is to add a new Extension decorator that accepts an object (of type Record<string, any>) which contains any extra properties that need to be inserted into the generated route.

## Breaking change?

No

### PR DESCRIPTION
**Pull request by molinab297-unisys**
### All Submissions:

- [x] Have you followed the guidelines in our [Contributing](https://github.com/lukeautry/tsoa/tree/master/docs/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/lukeautry/tsoa/pulls) for the same update/change?
- [x] Have you written unit tests?
- [x] Have you written unit tests that cover the negative cases (i.e.: if bad data is submitted, does the library respond properly)?
- [x] This PR is associated with an existing issue?

**Closing issues**

Closes #696 , #347 , and #294

### If this is a new feature submission:

- [x] Has the issue had a maintainer respond to the issue and clarify that the feature is something that aligns with the [goals](https://github.com/lukeautry/tsoa#goal) and [philosophy](https://github.com/lukeautry/tsoa#philosophy) of the project?

**Potential Problems With The Approach**

This PR only supports vendor extensions on individual operations. The OpenAPI spec says that vendor extensions can be placed in many locations, such as in the info section, individual paths, and responses (see here for more info: https://swagger.io/docs/specification/openapi-extensions/).

**Test plan**

Happy path test cases: Good data results in the correct spec file being generated.

Unhappy path test cases: Providing an extension without being prefixed with a '-x' is not allowed according to the OpenAPI spec, therefore an error is expected to be thrown. I'm not sure if passing invalid data to the Extension decorator warrants tests because the compiler will prevent users from supplying invalid data (invalid meaning it doesn't adhere to the ExtensionType interface).

### CODE REVIEW

File: src/decorators/extension.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export function Extension(_name: string, _value: ExtensionType | ExtensionType[]): Function {
```
Comment (ivan-shaban) [inline src/decorators/extension.ts:1]: What the reason to use underscore in variable names?

Comment (WoH) [reply src/decorators/extension.ts:1]: To indicate that they're not used.

File: tests/unit/swagger/schemaDetails.spec.ts [hunk 2]
```diff
@@ -127,4 +127,31 @@ describe('Schema details generation', () => {
       expect(errToTest!.message).to.eq(`Enums can only have string or number values, but enum MixedStringAndNumberEnum had number,string`);
     });
   });
+
+  describe('Extensions schema generation', () => {
```
Comment (WoH) [inline tests/unit/swagger/schemaDetails.spec.ts:131]: Can you please duplicate these tests to make sure this works for OpenAPI 3 aswell?

Comment (molinab297-unisys) [reply tests/unit/swagger/schemaDetails.spec.ts:131]: @WoH Done.

**Comment by molinab297-unisys** [comment]
I somehow messed up merging master into this branch, so I had to revert some commits and force push the fixes. Also I noticed that some changes got into master that are causing TSLint errors. It wouldn't let me commit with the errors present, so I just temporarily suppressed them (see my last commit).

**Comment by WoH** [comment]
LGTM, thank youvery much for adding this!

**Pull request merged**